src/main/java/org/opensearch/sdk/BaseExtension.java

```diff
@@ -66,14 +71,12 @@ public ExtensionSettings getExtensionSettings() {
         return this.settings;
     }
 
-    /**
-     * Returns components added by this extension.
-     *
-     * @param client A client to make requests to the system
-     * @param clusterService A service to allow watching and updating cluster state
-     * @param threadPool A service to allow retrieving an executor to run an async action
-     * @return A collection of objects
-     */
+    @Override
+    public void setExtensionsRunner(ExtensionsRunner extensionsRunner) {
```

Comment:
Any use case/reason to have this setter? Shouldn't we pass the object directly [like](https://github.com/opensearch-project/anomaly-detection/blob/feature/extensions/src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java#L112)

Reply:
> Any use case/reason to have this setter? Shouldn't we pass the object directly [like](https://github.com/opensearch-project/anomaly-detection/blob/feature/extensions/src/main/java/org/opensearch/ad/AnomalyDetectorExtension.java#L112)

Note in that example we're in a static `main()` method.  So while you could receive an `ExtensionsRunner` instance as a return value from the `run()` call, we would still have to operate on that new Extension instance somehow to pass the runner into it.

There's no guarantee that `main()` method will always be in the same class as the `Extension`, so setting a static variable in that class doesn't work.  You could have a standalone class with only the `main()` method that launches several extensions in the same JVM.

Reply:
More specifically here's an example that would theoretically work:
```java
/**
   * Runs the specified extension.
   *
   * @param extension The extension to run.
   * @return the ExtensionsRunner instance running the extension.
   * @throws IOException on failure to bind ports.
   */
  public static ExtensionsRunner run(Extension extension) throws IOException {
    logger.info("Starting extension " + extension.getExtensionSettings().getExtensionName());
    ExtensionsRunner runner = new ExtensionsRunner(extension);
    // initialize the transport service
    NettyTransport nettyTransport = new NettyTransport(runner);
    Settings runnerSettings = runner.getSettings();
    ThreadPool threadPool = new ThreadPool(runnerSettings);
    runner.extensionTransportService = nettyTransport.initializeExtensionTransportService(runnerSettings,
        threadPool);
    runner.startActionListener(0);
    return runner;
  }
```
and
```java
class HelloWorldExtension extends BaseExtension {
    private static ExtensionsRunner extensionsRunner;

    // ...

    public static void main(String[] args) throws IOException {
        // Execute this extension by instantiating it and passing to ExtensionsRunner
        runner = ExtensionsRunner.run(new HelloWorldExtension());
    }
}
```
But it's super-code-smelly setting a static variable after the class has been instantiated.  And as above, `static` also assumes the class executing `main()` is the same as the extension class, which is not a requirement.

Reply:
Okay so `main` is in the HelloWorldExtension.java here we don't have to make any changes to Hello World Extension
https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java#L62

Reply:
> Okay so `main` is in the HelloWorldExtension.java [here](https://github.com/opensearch-project/opensearch-sdk-java/blob/main/src/main/java/org/opensearch/sdk/sample/helloworld/HelloWorldExtension.java#L62) 

Yes but it doesn't have to be.

> we don't have to make any changes to Hello World Extension

We may need to pass the `runner` as an argument to a Rest Handler as in [this](https://github.com/opensearch-project/anomaly-detection/pull/725/files#diff-9d4fba791ae69d5fc024218799521d5a87e6c3cac867e7fbd2754a1d4abbb39f)

Reply:
Should we change the HelloWorldExtension.java then?

Reply:
No, this is going to be one of many things available in the Base Extension that may need to be passed. SDK Client is another. CreateComponents will probably produce even more.